### PR TITLE
Fixes 2975,2976: Add templates table/create modal

### PR DESCRIPTION
**Pull request by Andrewgdewar**
## Summary

Main Changes: 

- Added Templates table main list page ([HMS-2975](https://issues.redhat.com/browse/HMS-2975))
- Added ability to create templates with multi-page modal ([HMS-2976](https://issues.redhat.com/browse/HMS-2976)) 
- Added template list item "Delete" menu option (currently no confirmation modal).
- Added template list item "Edit"  menu option and accompanying modal (most functionality similar to create, with the caveat that arch/version cannot be changed). 

Additional Changes (that may effect tests/QE/documentation)
Rewrote/reorganized all routing: 
- The "repositories" page now lives at content/repositories
- The "add/edit/delete" repositories modals have had their routes simplified as a result to simple "add/edit/delete"
- The templates page lives at content/templates
- Template related modals are at the expected add/edit routes for them following the new pattern for content above.

## Caveat; we are currently waiting on feedback for the following items:
- Although there are some designs associated with the tickets, there are gaps in design, most notable the "Set up date" step within the edit/add template modal as well as for the main list page.
- There is currently some uncertainty as to which columns will appear on some of the add/edit modal steps, as well as on the main list page. 

## Testing steps

- As these are brand new pages, functionality/copy both need a thorough review.
- A general code review can be undertaken, as it is likely that functionality will remain even if content (columns/copy) is updated later.

### CODE REVIEW

**Comment by jlsherrill** [comment]
https://issues.redhat.com/browse/HMS-2975

**Comment by jlsherrill** [comment]
https://issues.redhat.com/browse/HMS-Fixes

**Comment by jlsherrill** [comment]
⚠️ This task currently requires qe-approval, but this PR does not fully resolve the task.  Please contact QE to determine appropriate testing required.

**Comment by jlsherrill** [comment]
the proper title is:  `Fixes 2975,2976:`  not sure why the pr validation isn't catching it, i'll look into it

**Comment by jlsherrill** [comment]
fixed here; https://github.com/content-services/content-sources-frontend/pull/194

**Comment by jlsherrill** [comment]
https://issues.redhat.com/browse/HMS-2976

**Comment by jlsherrill** [comment]
default urls with this PR: https://github.com/content-services/content-sources-backend/pull/538:

x86_64, rhel8:
        "url": "https://cdn.redhat.com/content/dist/rhel8/8/x86_64/appstream/os",
        "url": "https://cdn.redhat.com/content/dist/rhel8/8/x86_64/baseos/os",

x86_64, rhel9:
        "url": "https://cdn.redhat.com/content/dist/rhel9/9/x86_64/appstream/os",
        "url": "https://cdn.redhat.com/content/dist/rhel9/9/x86_64/baseos/os",

aarch64, rhel8:
        "url": "https://cdn.redhat.com/content/dist/rhel8/8/aarch64/appstream/os",
        "url": "https://cdn.redhat.com/content/dist/rhel8/8/aarch64/baseos/os",

aarch64, rhel9:
        "url": "https://cdn.redhat.com/content/dist/rhel9/9/aarch64/appstream/os",
        "url": "https://cdn.redhat.com/content/dist/rhel9/9/aarch64/baseos/os",

**Comment by xbhouse** [comment]
Screenshot 2024-02-01 at 2 05 25 PM

looks like the lorem ipsum words are still here in the set up date step of the create and edit modals

**Comment by jlsherrill** [comment]
For this warning:
image

*  is it possible to show the date of the selected snapshot in the table?  
* Are the package/errata count of the snapshot date or just of the introspected data?
* let me get you better text for that toggle

File: src/services/Templates/TemplateApi.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+export const deleteTemplateItem: (uuid: string) => Promise<void> = async (uuid: string) => {
+  const { data } = await axios.delete(`/api/content-sources/v1/templates/${uuid}`);
+  return data;
+};
+
+export const createTemplate: (request: TemplateRequest) => Promise<TemplateItem> = async (
+  request,
+) => {
+  const { data } = await axios.post('/api/content-sources/v1.0/templates/', request);
```
Comment (xbhouse) [inline src/services/Templates/TemplateApi.ts:86]: why is this endpoint v1.0? can we use either v1 or v1.0?

File: src/Pages/TemplatesTable/components/TemplateFilters.tsx [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+      distribution_arches.forEach((arch) => (arches[arch.name] = arch.label));
+      distribution_versions.forEach((version) => (versions[version.name] = version.label));
+      setVersionNamesLabels(versions);
+      setArchNamesLabels(arches);
+    }
+  }, [distribution_arches, distribution_versions]);
+
+  const Filter = useMemo(() => {
+    switch (filterType) {
+      case 'Search':
```
Comment (xbhouse) [inline src/Pages/TemplatesTable/components/TemplateFilters.tsx:128]: should this be 'Name' ? it's unclear from 'Search' what it's filtering on, but i think it's by name

**Comment by xbhouse** [comment]
Screenshot 2024-02-02 at 10 45 23 AM

in the review step of the create / edit modals, do you think it makes sense to remind the user that this date means `include snapshots up to this date` ?

**Comment by xbhouse** [comment]
Screenshot 2024-02-02 at 11 02 22 AM

do you think it makes sense to add input validation on the template name input if creating a template with a name that already exists (like in the repo create modal)? i do get an error that the name exists when hitting create and the template modal doesn't exit so maybe that's enough

**Comment by xbhouse** [comment]
Screenshot 2024-02-02 at 11 28 16 AM

if i delete the last template on the second page, i see this and can't see the other templates (even though it says i'm on the first page). i do see the first page of templates again if i just update the number of templates shown per page

File: src/services/Templates/TemplateApi.ts (hunk 1 above)
Comment (Andrewgdewar) [reply src/services/Templates/TemplateApi.ts:86]: I believe either works yup.

File: src/Pages/TemplatesTable/components/TemplateFilters.tsx (hunk 2 above)
Comment (Andrewgdewar) [reply src/Pages/TemplatesTable/components/TemplateFilters.tsx:128]: Hmmm I was unsure myself if it was on both Name/Description or just name.

**Comment by Andrewgdewar** [comment]
> Screenshot 2024-02-02 at 11 28 16 AM
> 
> if i delete the last template on the second page, i see this and can't see the other templates (even though it says i'm on the first page). i do see the first page of templates again if i just update the number of templates shown per page

Ahh yes, I forgot to implement the fix for that scenario, good catch.

**Comment by Andrewgdewar** [comment]
> Screenshot 2024-02-02 at 11 02 22 AM
> 
> do you think it makes sense to add input validation on the template name input if creating a template with a name that already exists (like in the repo create modal)? i do get an error that the name exists when hitting create and the template modal doesn't exit so maybe that's enough

I think it would be ideal, but we would have to build an API that can do that.

For now, hitting an error and having to go back a couple pages to adjust the title seems reasonable to me.

File: src/Pages/TemplatesTable/components/TemplateFilters.tsx (hunk 2 above)
Comment (xbhouse) [reply src/Pages/TemplatesTable/components/TemplateFilters.tsx:128]: all of my template descriptions have the word `description`, searching for `d` doesn't find them so it seems it's by name. i see the same Filter by Search for repositories in the add / edit template modals too, i think maybe those should be Filter by Name/URL?

File: src/Pages/TemplatesTable/components/AddTemplate/steps/RedhatRepositoriesStep.tsx [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+
+  const onSetPage = (_, newPage: number) => setPage(newPage);
+  const onPerPageSelect = (_, newPerPage: number, newPage: number) => {
+    setPerPage(newPerPage);
+    setPage(newPage);
+  };
+
+  const columnHeaders = ['Name', /* 'Label',*/ 'Advisories', 'Packages'];
+
+  const columnSortAttributes = ['name', '', '', 'package_count'];
```
Comment (xbhouse) [inline src/Pages/TemplatesTable/components/AddTemplate/steps/RedhatRepositoriesStep.tsx:93]: should both the name and packages column be sortable for the RH repositories step? looks like it's only sortable by name

File: src/Pages/TemplatesTable/components/TemplateFilters.tsx (hunk 2 above)
Comment (Andrewgdewar) [reply src/Pages/TemplatesTable/components/TemplateFilters.tsx:128]: Good catch thanks!

**Comment by xbhouse** [comment]
looks good!!

**Review by xbhouse — APPROVED** [review]

**Comment by jlsherrill** [comment]
for What does this mean, only in the error situation:

* No snapshots exist for these repositories on the specified date
* The closest snapshots after that date will be used
* Depending on the repository and time difference, this could cause a dependency issue

**Comment by jlsherrill** [comment]
I did find a 🐛   

1.  go to create a template
2.  select el9
3. select next
4. go back
5. select el8
6. select next

You'll still see el9 repos in the list.  Arch doesn't appear to have this bug

**Comment by Andrewgdewar** [comment]
> I did find a 🐛
> 
> 1. go to create a template
> 2. select el9
> 3. select next
> 4. go back
> 5. select el8
> 6. select next
> 
> You'll still see el9 repos in the list. Arch doesn't appear to have this bug

Fixed, good catch @jlsherrill!

File: src/Pages/TemplatesTable/components/AddTemplate/steps/DefineContentStep.tsx [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+  const allowedDistributionVersions = distribution_versions
+    .filter(({ label }) => ['8', '9'].includes(label))
+    .map(({ label, name }) => ({ value: label, label: name }));
+
+  return (
+    <Grid hasGutter>
+      <Title headingLevel='h1'>Define template content</Title>
+      <Text component={TextVariants.h6}>
+        Templates provide you with consistent content across environments and time by allowing you
```
Comment (swadeley) [inline src/Pages/TemplatesTable/components/AddTemplate/steps/DefineContentStep.tsx:55]:
```suggestion
        Templates provide consistent content across environments and time by enabling you
```

File: src/Pages/TemplatesTable/components/AddTemplate/steps/DefineContentStep.tsx [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        control the scope of package and advisory updates to be installed on selected systems.
```
Comment (swadeley) [inline src/Pages/TemplatesTable/components/AddTemplate/steps/DefineContentStep.tsx:56]:
```suggestion
        to control the scope of package and advisory updates to be installed on selected systems.
```

File: src/Pages/TemplatesTable/components/AddTemplate/steps/DefineContentStep.tsx [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+      </FormGroup>
+      <ExpandableSection
+        toggleText='What does it mean?'
+        aria-label='quickStart-expansion'
+        data-ouia-component-id='quickstart_expand'
+      >
+        <TextContent>
+          <TextList>
+            <TextListItem>
+              Configures clients to use date based snapshots of Red Hat and Custom repositories.
```
Comment (swadeley) [inline src/Pages/TemplatesTable/components/AddTemplate/steps/DefineContentStep.tsx:104]:
```suggestion
              Configures clients to use date-based snapshots of Red Hat and Custom repositories.
```

File: src/Pages/TemplatesTable/components/AddTemplate/steps/RedhatRepositoriesStep.tsx [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+          </Hide>
+        </Flex>
+      </Hide>
+      {showLoader ? (
+        <Bullseye data-ouia-safe={!isFetching} data-ouia-component-id='redhat_repositories_table'>
+          <EmptyTableState
+            notFiltered={searchQuery === ''}
+            clearFilters={() => setSearchQuery('')}
+            itemName='red hat repositories'
+            notFilteredBody='No Red Hat repositories match your desired version and arch'
```
Comment (swadeley) [inline src/Pages/TemplatesTable/components/AddTemplate/steps/RedhatRepositoriesStep.tsx:219]:
```suggestion
            notFilteredBody='No Red Hat repositories match the version and arch'
```
Style Guide does not like "desire" in tech docs.

File: src/Pages/TemplatesTable/components/AddTemplate/steps/SetUpDateStep.tsx [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+      '',
+      ContentOrigin.ALL,
+    );
+
+  return (
+    <Grid hasGutter>
+      <Title headingLevel='h1'>Set up date</Title>
+      <Text component={TextVariants.h6}>
+        This will include snapshots up to a specific date. Content of the snapshots created after
+        selected date will be displayed as applicable, not installable.
```
Comment (swadeley) [inline src/Pages/TemplatesTable/components/AddTemplate/steps/SetUpDateStep.tsx:105]:
```suggestion
        the selected date will be displayed as applicable, not installable.
```

**Comment by swadeley** [comment]
Hi, can you hide this text when no repos match:
"Select additional Red Hat repositories. Core repositories of your OS version have already been added."

Also, in this line "No red hat repositories", displayed just above, s/red hat/Red Hat/.
(I cannot find that in the diff)

**Comment by swadeley** [comment]
Hi, I have three repos:

image

"All" is highlighted. I get to the review stage and it shows no Custom Repositories are selected.

image

File: src/Pages/TemplatesTable/components/AddTemplate/steps/ReviewStep.tsx [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+      expanded.add(index);
+    }
+    setExpanded(new Set(expanded));
+  };
+
+  return (
+    <Grid hasGutter>
+      <Title headingLevel='h1'>Review</Title>
+      <Text component={TextVariants.h6}>
+        Review information and then click <b>{isEdit ? 'Confirm changes' : 'Create'}</b>.
```
Comment (swadeley) [inline src/Pages/TemplatesTable/components/AddTemplate/steps/ReviewStep.tsx:61]:
```suggestion
        Review the information and then click <b>{isEdit ? 'Confirm changes' : 'Create'}</b>.
```

**Comment by swadeley** [comment]
> Hi, can you hide this text when no repos match: "Select additional Red Hat repositories. Core repositories of your OS version have already been added."
> 
> Also, in this line "No red hat repositories", displayed just above, s/red hat/Red Hat/. (I cannot find that in the diff)

Hi @Andrewgdewar ,
Thank you for correcting the title case (Red Hat):
No Red Hat repositories match the version and arch

But now, notice below there is no space between the sentences:

`You can select additional Red Hat repositories.Core repositories of your OS version have been added.`

The new code to hide the text when there are no RHEL repo available is not detecting an available RHEL repo before I select it.

**Comment by swadeley** [comment]
lgtm, thank you

**Pull request merged**